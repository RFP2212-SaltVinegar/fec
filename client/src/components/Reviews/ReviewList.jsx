import React, {useState, useEffect} from 'react';
import axios from 'axios';
import SortOption from './SortOption.jsx';
import ReviewTile from './ReviewTile.jsx';
import RatingBreakdownSection from './RatingBreakdownSection/RatingBreakdownSection.jsx';
import WriteReviewModal from './WriteReviewModal.jsx';
import Search from './Search.jsx';
import '../../css/Reviews/listStyles.css'

const ReviewList = ({product_id, productName}) => {

  const [reviews, setReviews] = useState([]);
  const [reviewsOnPage, setReviewsOnPage] = useState([]);
  const [ratingData, setRatingData] = useState({});
  const [writeReview, setWriteReview] = useState(false);
  const [filters, setFilters] = useState([]);
  const [searchQuery, setSearchQuery] = useState('');
  const [charIds, setCharIds] = useState([]);

  useEffect(() => {
    callReviewData();
    callProductData();
  }, [filters, searchQuery]);

  const callReviewData = (sortChoice) => {
    if (product_id) {
      axios.get('http://localhost:3000/api/reviews/', {
        params: {
          "product_id": product_id,
          "sort": sortChoice || 'relevant',
          "count": 20}})
      .then((res) => {
        let newReviews = []
        if (searchQuery.length) {
          for (let review of res.data.results) {
            if (review.body.toLowerCase().includes(searchQuery.toLowerCase()) || review.summary.toLowerCase().includes(searchQuery.toLowerCase())) {
              newReviews.push(review);
            }
          }
        } else {
          newReviews = res.data.results;
        }
        if (filters.length) {
          let newFilteredReviews = [];
          for (let review of newReviews) {
            if (filters.includes(review.rating)) {
              newFilteredReviews.push(review);
            }
          }
          setReviews(newFilteredReviews);
          setReviewsOnPage(newFilteredReviews.slice(0, 2));
        } else {
          setReviews(newReviews);
          setReviewsOnPage(newReviews.slice(0, 2));
        }
      })
      .catch((err) => {
        console.log(err);
      })
    }
  }

  const callProductData = () => {
    if (product_id) {
      axios.get('http://localhost:3000/api/reviews/meta', {params: {"product_id": product_id}})
      .then((res) => {
        setRatingData(res.data);
        let newCharacteristics = [];
        for (let i = 0; i < Object.values(res.data.characteristics).length; i++) {
          newCharacteristics.push(Object.values(res.data.characteristics)[i].id);
        }
        setCharIds(newCharacteristics);
      })
      .catch((err) => {
        console.log(err);
      })
    }
  }

  const handleSortChange = (option) => {
    callReviewData(option);
  }

  const showMore = (e) => {
    e.preventDefault();
    let idx = reviewsOnPage.length;
    let newReviewsonPage = reviewsOnPage.slice();
    if ((idx + 2) > reviews.length) {
      newReviewsonPage.push(reviews[idx]);
    } else {
      newReviewsonPage.push(reviews[idx]);
      idx++;
      newReviewsonPage.push(reviews[idx]);
    }
    setReviewsOnPage(newReviewsonPage);
  }

  const handleFilterRL = (index) => {
    if (typeof index === 'object') {
      setFilters([]);
    } else if (filters.includes(index)) {
      let newFilters = filters.slice();
      for (let i = 0; i < newFilters.length; i++) {
        if (newFilters[i] === index) {
          newFilters.splice(i, 1);
        }
      }
      setFilters(newFilters);
    } else {
      let newFilters = filters.slice();
      newFilters.push(index);
      setFilters(newFilters);
    }
  }

  const getSearchQuery = (query) => {
    if (query.length >= 3) {
      setSearchQuery(query);
    } else {
      setSearchQuery('')
    }
  }

  const openWriteReview = () => {
    setWriteReview(true)
    document.body.style.overflowY = 'hidden';
  }

  return (
    <>
      <div className='left-reviews' id="review">
        <RatingBreakdownSection ratingData={ratingData} handleFilter={handleFilterRL} filters={filters}/>
        <button onClick={openWriteReview} >Write Review</button>
        {writeReview
      ? <WriteReviewModal setWriteReview={setWriteReview} productName={productName} charIds={charIds} characteristics={ratingData.characteristics} product_id={product_id}/>
      : null}
      </div>
      <div className='right-reviews'>
        <div className='right-reviews-header'>
          <SortOption handleSortChange={handleSortChange} />
          <Search getQuery={getSearchQuery}/>
        </div>
        <div className='right-reviews-cards'>
          {Object.keys(reviews).length
          ? <>
              {reviewsOnPage.map((review, idx) =>
                <ReviewTile review={review} key={idx} />
              )}
            </>
          : null
          }
        </div>
        {(reviews.length >= 2 && reviews.length !== reviewsOnPage.length)
          ? <button onClick={showMore}>More Reviews</button>
          : null}
      </div>
      <br/>
    </div>
  );
}

export default ReviewList;