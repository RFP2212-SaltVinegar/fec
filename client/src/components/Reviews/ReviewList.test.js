import renderer from 'react-test-renderer';
import ReviewList from './ReviewList.jsx';
import {render, fireEvent, screen, waitFor, getByText} from '@testing-library/react';
import '@testing-library/jest-dom'

it('renders properly', () => {
  const component = renderer.create(
    <ReviewList />
  );

  let test = component.toJSON();
  expect(test).toMatchSnapshot();
});

it('does not search when less than 3 characters are typed in the search bar', async () => {
  render(<ReviewList product_id={40344} productName={'funny hat'}/>);
  return waitFor(() => expect(screen.queryByText('Write Review')).toBeInTheDocument())
    .then(() => {
      const inputNode = screen.getByPlaceholderText('Search...')
      fireEvent.keyDown(inputNode, {Key: 'A', code: 'KeyA'})
      console.log(inputNode)
    })
    .catch((err) => {
      console.log('error', err);
    })
});

it('adds filter to page when progress bar is clicked', () => {
  render(<ReviewList product_id={40344} productName={'funny hat'}/>);
  return waitFor(() => expect(screen.queryByText('Write Review')).toBeInTheDocument())
    .then(() => {

    })
    .catch((err) => {
      console.log('error', err);
    })
})

it('opens write a review modal on button click', () => {
  render(<ReviewList product_id={40352} productName={'funny hat'}/>);
  // const button = getByText('Write Review')
  return waitFor(() => expect(screen.queryByText('Write Review')).toBeInTheDocument())
    .then(()=> {
      fireEvent.click(screen.getByText('Write Review'));
    })

})