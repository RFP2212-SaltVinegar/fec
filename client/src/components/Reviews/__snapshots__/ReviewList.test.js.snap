// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders properly 1`] = `
[
  <div
    className="left-reviews"
    id="review"
  >
    <button
      onClick={[Function]}
    >
      Write Review
    </button>
  </div>,
  <div
    className="right-reviews"
  >
    <form
      className="sort"
    >
      <label>
        Sort on: 
        <select
          onChange={[Function]}
          value="relevant"
        >
          <option
            value="helpful"
          >
            Helpful
          </option>
          <option
            value="newest"
          >
            Newest
          </option>
          <option
            value="relevant"
          >
            Relevant
          </option>
        </select>
      </label>
    </form>
    <input
      onChange={[Function]}
      placeholder="Search..."
      type="text"
    />
  </div>,
  <br />,
]
`;
